eta/core/datasets.py

```diff
@@ -1338,3 +1373,839 @@ def required(cls):
+                        or frame_no >= record.clip_end_frame):
+                    continue
+
+                for detected_object in frame.objects:
+                    if detected_object.label != self.object_label:
+                        continue
+
+                    for attr in detected_object.attrs:
+                        if attr.name == self.attr_name:
+                            obj_idx = (
+                                detected_object.index
+                                if detected_object.index is not None
+                                else NO_ID
+                            )
+
+                            helper_dict[obj_idx].add(attr.value)
+
+                            break
+
+            # At this point, the keys of helper dict are unique
+            # object indices for objects of type self.object_label.
+            # The values are unique attribute values for self.attr_name.
+
+            if len(helper_dict):
+                helper = (i, [])
+                for s in helper_dict.values():
+                    helper[1].extend(s)
+                helper_list.append(helper)
+
+        return helper_list
+
+    def _get_target_count(self, counts):
+        '''Compute the target count that we would like to balance each value to.
+
+        Args:
+            A: the occurrence matrix computed in Balancer._get_occurrence_matrix
+
+        Returns: Integer target value
+        '''
+        if self.target_count:
+            return self.target_count
+        return int(np.quantile(counts, self.target_quantile))
+
+    def _get_keep_idxs(self, A, counts, target_count):
+        '''Algorithm fun! This function chooses the set of records to keep (and
+        remove).
+
+        There's still plenty of potential for testing and improvement here.
+
+
+        This problem can be posed as:
+            minimize:
+                ||np.dot(A, x) - b||
+            subject to:
+                x[i] is an element of [0, 1]
+
+        and different algorithms may be substituted in.
+
+        Args:
+            A: the occurrence matrix computed in Balancer._get_occurrence_matrix
+            counts: (vector) the original counts for each value
+            target_count: (int) the target value
+
+        Returns:
+            a list of integer indices to keep
+        '''
+        b = counts - target_count
+
+        x = self._greedy(A, b)
+
+        return np.where(x==0)[0]
+
+    def _random(self, A, b):
+        '''A random search algorithm for finding the indices to omit.
+
+        Args:
+            A: the occurrence matrix
+            b: the target vector to match
+
+        Returns:
+            x: the solution vector. x[j]=1 -> omit the j'th record
+        '''
+        best_x = np.zeros(A.shape[1], dtype=np.dtype('int'))
+        best_score = self._solution_score(b - np.dot(A, best_x))
+
+        import random
+        random.seed(1)
+        for _ in range(10000):
```

Comment:
always performing 10K O(n^2) operations? yikes

I haven't translated what you're doing into math, but this can probably be cast as a linear program and solved using something off-the-shelf like `scipy.optimize.linprog`
https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.linprog.html

Reply:
This is a very simple and bad solution which can easily be exchanged with `_greedy` but is not being used right now. The purpose was that multiple algorithms can be easily swapped in and out here and this one, while bad, is easy to understand and stands as an example.

10K is totally arbitrary, and also, I believe this is O(NM) where: N - number of samples, M - number of attribute values and it should be safe to assume that in most cases N >> M.